src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -177,13 +184,73 @@ private void setupTemplateAndPush(Repository repository, Resource[] resources, S
             fileTargets.add("${exerciseName}");
             fileReplacements.add(programmingExercise.getTitle());
 
-            //TODO: for some reason, this code does not replace the elements in the file test.json
-
             fileService.replaceVariablesInFileRecursive(repository.getLocalPath().toAbsolutePath().toString(), fileTargets, fileReplacements);
 
             gitService.stageAllChanges(repository);
             gitService.commitAndPush(repository, templateName + "-Template pushed by ArTEMiS");
             repository.setFiles(null); // Clear cache to avoid multiple commits when ArTEMiS server is not restarted between attempts
         }
     }
+
+    /**
+     * This method calls the StructureOracleGenerator, generates the string out of the JSON representation of the structure
+     * oracle of the programming exercise and returns true if the file was updated or generated, false otherwise.
+     * This can happen if the contents of the file have not changed.
+     * @param solutionRepoURL The URL of the solution repository.
+     * @param exerciseRepoURL The URL of the exercise repository.
+     * @param testRepoURL The URL of the tests repository.
+     * @param testsPath The path to the tests folder, e.g. the path inside the repository where the structure oracle file will be saved in.
+     * @return True, if the structure oracle was successfully generated or updated, false if no changes to the file were made.
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public boolean generateStructureOracleFile(URL solutionRepoURL, URL exerciseRepoURL, URL testRepoURL, String testsPath) throws IOException, InterruptedException {
+        Repository solutionRepository = gitService.getOrCheckoutRepository(solutionRepoURL);
+        Repository exerciseRepository = gitService.getOrCheckoutRepository(exerciseRepoURL);
+        Repository testRepository = gitService.getOrCheckoutRepository(testRepoURL);
+
+        gitService.pull(solutionRepository);
+        gitService.pull(exerciseRepository);
+        gitService.pull(testRepository);
```

Comment:
This code is invoked by instructors when creating programming exercises and it currently takes less than 5s so I don't think it is necessary to execute it in parallel. In other cases this might make more sense.

Reply:
Yes, even the whole oracle process generation together with the repo processes also take less than 5s, so there should be no problems.